### PR TITLE
[SPARK-16329] [SQL] Star Expansion over Table Containing No Column 

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

Star expansion over a table containing zero column does not work since 1.6. However, it works in Spark 1.5.1. This PR is to fix the issue in the master branch.

For example, 

``` scala
val rddNoCols = sqlContext.sparkContext.parallelize(1 to 10).map(_ => Row.empty)
val dfNoCols = sqlContext.createDataFrame(rddNoCols, StructType(Seq.empty))
dfNoCols.registerTempTable("temp_table_no_cols")
sqlContext.sql("select * from temp_table_no_cols").show
```

Without the fix, users will get the following the exception:

```
java.lang.IllegalArgumentException: requirement failed
        at scala.Predef$.require(Predef.scala:221)
        at org.apache.spark.sql.catalyst.analysis.UnresolvedStar.expand(unresolved.scala:199)
```
#### How was this patch tested?

Tests are added

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @rxin @cloud-fan @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 1]
```diff
@@ -215,6 +215,8 @@ abstract class Star extends LeafExpression with NamedExpression {
 case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevaluable {
 
   override def expand(input: LogicalPlan, resolver: Resolver): Seq[NamedExpression] = {
+    // When the table does not have any column, the input LogicalPlan does not have any column
+    if (input.output.isEmpty) return Seq.empty[NamedExpression]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: how about we fix the next code block? e.g.
```
if (target.isEmpty) {
  // If there is no table specified, use all input attributes.
  input.output
} else if (target.get.size == 1) {
  // If there is a table, pick out attributes that are part of this table.
  input.output.filter(_.qualifier.exists(resolver(_, t.head)))
} else {
  ...... // the rest
}
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: Sounds good to me. Let me change it. Thanks!

**Comment by maropu** [comment]
Great work! One question;
I got weird results from this fix.
```
scala> case class DATA(a: Int)
defined class DATA

scala> val df1 = Seq(DATA(1)).toDF
df1: org.apache.spark.sql.DataFrame = [a: int]

scala> val df2 = df1.drop($"a")
df2: org.apache.spark.sql.DataFrame = []

// OK
scala> df2.select($"*").show
++
||
++
||
++

// OK?
scala> df2.select($"hoge.*").show
++
||
++
||
++
```
Would be better to throw an easy-to-understand exception if unqualified inputs detected in this case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: @cloud-fan The latest changes fail a test case for star expansion in Struct. To resolve a star expansion for Stuct, the whole solution is using two steps when it has a one level qualifier. Try to resolve it as a table qualifier at first. If the returned attributes are empty, then try to resolve it as a struct qualifier.

**Comment by gatorsmile** [comment]
@maropu Sure, let me resolve it. Thanks for catching it!

**Comment by cloud-fan** [comment]
thanks, merging to master!

@gatorsmile can you backport it to 2.0 and fix the conflicts? thanks!

**Comment by cloud-fan** [comment]
and also 1.6 please.

**Comment by gatorsmile** [comment]
Sure, will do it soon. Thanks!

**Pull request closed without merging**